### PR TITLE
Acquire and persist user information on login.

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

UCM needs to know the user handle in order to construct branch and project names for share; 

This PR fetches user information as part of `auth.login` and stores it with your credentials.

## Implementation notes

As part of `auth.login` and any calls to `ensureAuthenticatedWithCodeserver` we fetch/refresh the user info using the `userInfoEndpoint` provided by the codeserver.

Migration:
This changes the credentials file format to require user info; I could add a bunch of code to migrate it properly from the old format to the new one, but honestly I think it's cleaner and simpler to just require the user to log in again the next time they try to do something, logging in is such a fast and painless thing I don't think it's worth complicating the source code to avoid it. So right now, if the credentials file fails to parse it just wipes it and replaces it with an empty one.

## Interesting/controversial decisions

Note that there's still no notion of a user's handle unless we're talking about a specific codeserver (e.g. Share); since the user could have a different handle on each of their "enterprise" codeservers.

If we need a default handle we may need the user to set their current preferred codeserver or something.

## Test coverage

- [x] test locally, integration test with the projects work.

## Loose ends

- [x] Need to deploy the Enlil side before merging this! https://github.com/unisoncomputing/enlil/pull/151

### CODE REVIEW

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 1]
```diff
@@ -1999,7 +1999,7 @@ handlePushToUnisonShare remote@WriteShareRemotePath {server, repo, path = remote
   let codeserver = Codeserver.resolveCodeserver server
   let baseURL = codeserverBaseURL codeserver
   let sharePath = Share.Path (shareUserHandleToText repo Nel.:| pathToSegments remotePath)
-  ensureAuthenticatedWithCodeserver codeserver
+  _userInfo <- ensureAuthenticatedWithCodeserver codeserver
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:2012]: Here's where you'd get the user's handle for creating projects/branches during a push.

**Review by mitchellwrosen — APPROVED** [review]

**Comment by aryairani** [comment]
What happens to the workflow if the user handle changes on the server after logging in?

**Comment by aryairani** [comment]
What happens to the workflow if the user handle changes on the server after logging in?

Should we instead have some way to construct branch and project names that can be atomic? (like the @me thing, and share decodes it?)

@ChrisPenner @mitchellwrosen @tstat

**Comment by ChrisPenner** [comment]
@aryairani  If your handle changes this will update it the next time you `auth.login`.

We could use UserIDs in the APIs instead if we like instead, but of course I think part of using paths and handles was meant to avoid mentioning users where possible.

The worst that could happen is that you get a 403 when trying to push a branch after changing your handle and you'd need to re-auth; although I'm not sure what would happen to any projects or branches you have in your UCM that mention the name of the old handle 🤔 

I think `@me` could be nice as well, I'd just want to think about it a little bit before making the jump 🤔

**Comment by mitchellwrosen** [comment]
> What happens to the workflow if the user handle changes on the server after logging in?

That should be fine; I suppose you mean "would a user lose access to all of their feature branches if they change their handle?" The answer is no: Share doesn't store branch names literally as "@user/branch", but rather the pair `(user id, branch name)`.

> Should we instead have some way to construct branch and project names that can be atomic? (like the @me thing, and share decodes it?)

I don't think we need to, or at least, I think we need to fully spec out the idea before considering it a viable alternative to the one we designed already that doesn't need to use a special `@me` sentinel.

**Comment by aryairani** [comment]
@mitchellwrosen (cc @ChrisPenner) Sorry, what I meant was what happens if
1. I auth.login
2. I change my username on the server
3. I do something in ucm that uses that cached user info (not sure exactly what use case might be problematic, what about  creating a contributor branch?)

Would that work, or could I end up stuck with an error, creating a branch for a non-existent user and not knowing what to do to fix it? Or what would happen?

It's okay in the short term for UCM to do something weird in this rare case, but I was worried that we were unnecessarily creating an "ask server for my info just to send it back to the same server, but possibly problematically out of date at this point" situation when the server knows my info to begin with.

What do you guys think?

**Comment by mitchellwrosen** [comment]
Ah, I understand now. 

You are suggesting we use a special syntax during remote branch creation that Share interprets as "from the calling user".

That seems fine. Note it would only affect first-time pushes of branches that aren't prefixed with a user slug, on users' machines that have some existing login session.

Already if a user chooses to manually namespace their contributor branches locally, as a way of keeping them separate from maintainer branches, they'll be unable to push (before a remote mapping is established) if they change their username.

**Comment by ChrisPenner** [comment]
Chatted with Arya in Slack, I think we're pretty safe to merge this for now, then can optionally add some sort of `@me` thing in the future if it comes up again 👍🏼

**Pull request merged**